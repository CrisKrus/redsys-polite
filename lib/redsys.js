module.exports =
/******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};

/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {

/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId])
/******/ 			return installedModules[moduleId].exports;

/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			exports: {},
/******/ 			id: moduleId,
/******/ 			loaded: false
/******/ 		};

/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);

/******/ 		// Flag the module as loaded
/******/ 		module.loaded = true;

/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}


/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;

/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;

/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";

/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(0);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ function(module, exports) {

	"use strict";

	Object.defineProperty(exports, "__esModule", {
	    value: true
	});


	var Redsys = function Redsys(options) {
	    this.terminal = options.terminal;
	    this.name = options.name;
	    this.merchant = options.merchantCode;
	    this.language = options.language;
	    this.url = options.url;
	    this.secret = options.secret;
	};

	var RedsysBuilder = function RedsysBuilder() {
	    this.name = "Default-Redsys";
	    this.terminal = "1";
	    this.language = "auto";
	    // Production URL
	    this.url = "https://sis.redsys.es/sis/realizarPago";
	    this.setMerchantCode = function (merchant_code) {
	        this.merchantCode = merchant_code;
	        return this;
	    };
	    this.setTerminal = function (terminal_number) {
	        this.terminal = terminal_number;
	        return this;
	    };
	    this.setName = function (name) {
	        this.name;
	        return this;
	    };
	    this.setTitular = function (titular) {
	        this.titular = titular;
	        return this;
	    };
	    this.setSecret = function (secret) {
	        this.secret = secret;
	        return this;
	    };
	    this.enableDebug = function () {
	        // Change to debug url
	        this.url = "https://sis-t.redsys.es:25443/sis/realizarPago";
	    };
	    this.build = function () {
	        if (this.merchantCode == undefined) throw new Error("Merchant Code not set");

	        if (this.titular == undefined) throw new Error("Titular not set");

	        if (this.secret == undefined) throw new Error("Secret not set");

	        return new Redsys(this);
	    };
	};

	exports.default = RedsysBuilder;
	module.exports = exports["default"];

/***/ }
/******/ ]);