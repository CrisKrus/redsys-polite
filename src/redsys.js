/*jshint esversion:6*/
import PaymentBuilder from './payment';
import crypto from 'crypto';

class Redsys {
  constructor(parmeters) {
    Object.assign(this, options);
    this.generateMerchantParams = function(payment) {
      return {
        "DS_MERCHANT_AMOUNT": payment.total,
        "DS_MERCHANT_ORDER": payment.order_id,
        "DS_MERCHANT_MERCHANTCODE": this.merchantCode,
        "DS_MERCHANT_CURRENCY": payment.currency,
        "DS_MERCHANT_TRANSACTIONTYPE": this.transaction_type,
        "DS_MERCHANT_TERMINAL": this.terminal,
        "DS_MERCHANT_MERCHANTURL": payment.redirect_urls.merchant_url,
        "DS_MERCHANT_URLOK": payment.redirect_urls.ok_url,
        "DS_MERCHANT_URLKO": payment.redirect_urls.cancel_url,
        'DS_MERCHANT_CONSUMERLANGUAGE': '001',
        'DS_MERCHANT_TITULAR': this.titular,
        'DS_MERCHANT_MERCHANTNAME': this.name,

        // Test code
        //"DS_MERCHANT_PAN":"4548812049400004",
        //"DS_MERCHANT_EXPIRYDATE":"1220",
        //"DS_MERCHANT_CVV2":"123"
      };
    };
  }

  encodeOrder(order_id, secret) {
    var secretKey = new Buffer(secret, 'base64');
    var iv = new Buffer(8);
    iv.fill(0);
    var cipher = crypto.createCipheriv('des-ede3-cbc', secretKey, iv);
    cipher.setAutoPadding(false);
    var res = cipher.update(order_id, 'utf8', 'base64') + cipher.final('base64');
    return res;
  }

  doSignature(order_encoded, merchantData) {
    var hexMac256 = crypto.createHmac("sha256", new Buffer(order_encoded, 'base64')).update(merchantData).digest("hex");
    return new Buffer(hexMac256, 'hex').toString('base64');
  }

  validateSignature(signature, merchantData){
    // Se decodifica la clave Base64
    let secretKey = new Buffer(secret, 'base64');
    // Se decodifican los datos Base64
    // $decodec = $this->base64_url_decode($datos);
    // Los datos decodificados se pasan al array de datos
    // $this->stringToArray($decodec);
    let decodedData = JSON.parse(new Buffer(merchantData,'base64'));

/*

// Se diversifica la clave con el Número de Pedido
$key = $this->encrypt_3DES($this->getOrderNotif(), $key);

// MAC256 del parámetro Ds_Parameters que envía Redsys
$res = $this->mac256($datos, $key);
// Se codifican los datos Base64
return $this->base64_url_encode($res);	*/
    var hexMac256 = crypto.createHmac("sha256", new Buffer(order_encoded, 'base64')).update(merchantData).digest("hex");
    return new Buffer(hexMac256, 'hex').toString('base64');

  }

  getFormData(payment) {
    // Merchant parameters as a Base64-JSON
    var merchant = new Buffer(JSON.stringify(this.generateMerchantParams(payment))).toString('base64');

    // Calculate signature
    var order_encoded = this.encodeOrder(payment.order_id, this.secret);
    var signature = this.doSignature(order_encoded, merchant);

    return {
      'redsys_url': this.url,
      'Ds_SignatureVersion': 'HMAC_SHA256_V1',
      'Ds_MerchantParameters': merchant,
      'Ds_Signature': signature,
    };
  }
}

class RedsysBuilder {
  constructor() {
    this.name = "Default-Redsys";
    this.terminal = "1";
    this.language = "auto";
    this.transaction_type = "0";
    // Production URL
    this.url = "https://sis.redsys.es/sis/realizarPago";
  }
  setMerchantCode(merchant_code) {
    this.merchantCode = merchant_code;
    return this;
  }
  setTerminal(terminal_number) {
    this.terminal = terminal_numbver;
    return this;
  }
  setName(name) {
    this.name = name;
    return this;
  }
  setTitular(titular) {
    this.titular = titular;
    return this;
  }
  setSecret(secret) {
    this.secret = secret;
    return this;
  }
  enableDebug() {
    // Change to debug url
    this.url = "https://sis-t.redsys.es:25443/sis/realizarPago";
    return this;
  }
  build() {
    if (this.merchantCode === undefined) throw new Error("Merchant Code not set");
    if (this.titular === undefined) throw new Error("Titular not set");
    if (this.secret === undefined) throw new Error("Secret not set");
    return new Redsys(this);
  }
}

export {
  PaymentBuilder,
  RedsysBuilder
};
