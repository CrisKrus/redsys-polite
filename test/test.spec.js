/*jshint esversion:6*/
require('dotenv').config()
import test from 'ava';
import {RedsysBuilder, PaymentBuilder} from '../lib/redsys';

const secret_data = {
  commerce_code: process.env.COMMERCE_CODE || '000000000',
  secret_code: process.env.COMMERCE_SECRET || "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
};
console.log(secret_data);

test('builder works with all required properties set',t => {
  var redsys = new RedsysBuilder()
              .setMerchantCode(secret_data.commerce_code)
              .setTitular("Marc Pomar")
              .setSecret(secret_data.secret_code)
              .build();
  t.is(redsys.secret, secret_data.secret_code);
  t.not(redsys.url,undefined);
});

test('throws error if missing properties on builder',t => {
  var builder = new RedsysBuilder()
      .setMerchantCode("33")
      .setTitular("Marc Pomar");
  t.throws(() => builder.build());
});

test('throws error if missing properties on builder',t => {
  var paymentBuilder = new PaymentBuilder()
      .setTotal(3.20)
      .setOrderId("123456")
      .setUrlMerchant("http://faable.com")
      .setUrlCancel("http://faable.com/cancel")
      .setUrlOK("http://faable.com/accept");
  t.notThrows(() => paymentBuilder.build());
});


test('throws error if missing properties on builder',t => {
  const redsys = new RedsysBuilder()
      .setMerchantCode(secret_data.commerce_code)
      .setName("Faable.com")
      .setTitular("Marc Pomar")
      .setSecret(secret_data.secret_code)
      .enableDebug()
      .build();

  const payment = new PaymentBuilder()
      .setTotal(3.20)
      .setOrderId("1")
      .setUrlMerchant("http://faable.com")
      .setUrlCancel("http://faable.com/cancel")
      .setUrlOK("http://faable.com/accept")
      .build();

  const form_encoded_params = redsys.getFormData(payment);
  const all_params = redsys.generateMerchantParams(payment);
  t.not(form_encoded_params, undefined);
  t.not(all_params, undefined);
  t.snapshot(form_encoded_params);
  t.snapshot(all_params);
});
