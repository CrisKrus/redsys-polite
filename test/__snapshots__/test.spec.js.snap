// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws error if missing properties on builder 1`] = `
Object {
  "Ds_MerchantParameters": "eyJEU19NRVJDSEFOVF9BTU9VTlQiOjMyMCwiRFNfTUVSQ0hBTlRfT1JERVIiOiIwMDAwMDAwMSIsIkRTX01FUkNIQU5UX01FUkNIQU5UQ09ERSI6IjAwMDAwMDAwMCIsIkRTX01FUkNIQU5UX0NVUlJFTkNZIjo5NzgsIkRTX01FUkNIQU5UX1RSQU5TQUNUSU9OVFlQRSI6IjAiLCJEU19NRVJDSEFOVF9URVJNSU5BTCI6IjEiLCJEU19NRVJDSEFOVF9NRVJDSEFOVFVSTCI6Imh0dHA6Ly9mYWFibGUuY29tIiwiRFNfTUVSQ0hBTlRfVVJMT0siOiJodHRwOi8vZmFhYmxlLmNvbS9hY2NlcHQiLCJEU19NRVJDSEFOVF9VUkxLTyI6Imh0dHA6Ly9mYWFibGUuY29tL2NhbmNlbCIsIkRTX01FUkNIQU5UX0NPTlNVTUVSTEFOR1VBR0UiOiIwMDEiLCJEU19NRVJDSEFOVF9USVRVTEFSIjoiTWFyYyBQb21hciIsIkRTX01FUkNIQU5UX01FUkNIQU5UTkFNRSI6IkZhYWJsZS5jb20ifQ==",
  "Ds_Signature": "yGPd0mTXadmMVP/2SzblwBK549F7nEF9EunDpBlV0qE=",
  "Ds_SignatureVersion": "HMAC_SHA256_V1",
  "redsys_url": "https://sis-t.redsys.es:25443/sis/realizarPago",
}
`;

exports[`throws error if missing properties on builder 2`] = `
Object {
  "DS_MERCHANT_AMOUNT": 320,
  "DS_MERCHANT_CONSUMERLANGUAGE": "001",
  "DS_MERCHANT_CURRENCY": 978,
  "DS_MERCHANT_MERCHANTCODE": "000000000",
  "DS_MERCHANT_MERCHANTNAME": "Faable.com",
  "DS_MERCHANT_MERCHANTURL": "http://faable.com",
  "DS_MERCHANT_ORDER": "00000001",
  "DS_MERCHANT_TERMINAL": "1",
  "DS_MERCHANT_TITULAR": "Marc Pomar",
  "DS_MERCHANT_TRANSACTIONTYPE": "0",
  "DS_MERCHANT_URLKO": "http://faable.com/cancel",
  "DS_MERCHANT_URLOK": "http://faable.com/accept",
}
`;
